### ISSUE TITLE
多GPU训练，triplet_loss一直为0，loss_cls显示正常

### ISSUE THREAD

**Issue by NiuCY** [JDAI-CV/fast-reid#432]
我的config：
CUDNN_BENCHMARK: true
DATALOADER:
  NAIVE_WAY: true
  NUM_INSTANCE: 16
  NUM_WORKERS: 8
  PK_SAMPLER: true
DATASETS:
  COMBINEALL: false
  NAMES:
  - VeRi
  TESTS:
  - VeRi
INPUT:
  AUGMIX_PROB: 0.0
  AUTOAUG_PROB: 0.1
  CJ:
    BRIGHTNESS: 0.15
    CONTRAST: 0.15
    ENABLED: false
    HUE: 0.1
    PROB: 0.5
    SATURATION: 0.1
  DO_AFFINE: false
  DO_AUGMIX: false
  DO_AUTOAUG: true
  DO_FLIP: true
  DO_PAD: true
  FLIP_PROB: 0.5
  PADDING: 10
  PADDING_MODE: constant
  REA:
    ENABLED: true
    PROB: 0.5
    VALUE:
    - 123.675
    - 116.28
    - 103.53
  RPT:
    ENABLED: false
    PROB: 0.5
  SIZE_TEST:
  - 128
  - 256
  SIZE_TRAIN:
  - 128
  - 256
KD:
  MODEL_CONFIG: ''
  MODEL_WEIGHTS: ''
MODEL:
  BACKBONE:
    DEPTH: 50x
    FEAT_DIM: 2048
    LAST_STRIDE: 1
    NAME: build_resnet_backbone
    NORM: BN
    PRETRAIN: true
    PRETRAIN_PATH: ''
    WITH_IBN: true
    WITH_NL: true
    WITH_SE: false
  DEVICE: cuda:0
  FREEZE_LAYERS:
  - backbone
  HEADS:
    CLS_LAYER: circleSoftmax
    EMBEDDING_DIM: 0
    MARGIN: 0.35
    NAME: EmbeddingHead
    NECK_FEAT: after
    NORM: BN
    NUM_CLASSES: 3155
    POOL_LAYER: gempoolP
    SCALE: 64
    WITH_BNNECK: true
  LOSSES:
    CE:
      ALPHA: 0.2
      EPSILON: 0.1
      SCALE: 1.0
    CIRCLE:
      GAMMA: 128
      MARGIN: 0.25
      SCALE: 1.0
    COSFACE:
      GAMMA: 128
      MARGIN: 0.25
      SCALE: 1.0
    FL:
      ALPHA: 0.25
      GAMMA: 2
      SCALE: 1.0
    NAME:
    - CrossEntropyLoss
    - TripletLoss
    TRI:
      HARD_MINING: true
      MARGIN: 0.0
      NORM_FEAT: false
      SCALE: 1.0
  META_ARCHITECTURE: Baseline
  PIXEL_MEAN:
  - 123.675
  - 116.28
  - 103.53
  PIXEL_STD:
  - 58.395
  - 57.120000000000005
  - 57.375
  QUEUE_SIZE: 8192
  WEIGHTS: logs/veri/resent50_ibn_ver/model_best.pth
OUTPUT_DIR: logs/veri/resent50_ibn_ver
SOLVER:
  BASE_LR: 0.01
  BIAS_LR_FACTOR: 1.0
  CHECKPOINT_PERIOD: 20
  DELAY_EPOCHS: 30
  ETA_MIN_LR: 7.7e-05
  FP16_ENABLED: false
  FREEZE_FC_ITERS: 0
  FREEZE_ITERS: 10
  GAMMA: 0.1
  HEADS_LR_FACTOR: 1.0
  IMS_PER_BATCH: 64
  MAX_EPOCH: 240
  MOMENTUM: 0.9
  NESTEROV: true
  OPT: SGD
  SCHED: CosineAnnealingLR
  STEPS:
  - 40
  - 90
  WARMUP_EPOCHS: 2000
  WARMUP_FACTOR: 0.1
  WARMUP_METHOD: linear
  WEIGHT_DECAY: 0.0005
  WEIGHT_DECAY_BIAS: 0.0005
TEST:
  AQE:
    ALPHA: 3.0
    ENABLED: false
    QE_K: 5
    QE_TIME: 1
  EVAL_PERIOD: 20
  FLIP_ENABLED: false
  IMS_PER_BATCH: 128
  METRIC: cosine
  PRECISE_BN:
    DATASET: Market1501
    ENABLED: false
    NUM_ITER: 300
  RERANK:
    ENABLED: false
    K1: 20
    K2: 6
    LAMBDA: 0.3
  ROC_ENABLED: false

我执行训练的命令：python tools/train_net.py  --config-file logs/veri/dataset0304/config.yaml --num-gpus 4

**Comment by NiuCY** [comment]
我在单卡训练的时候，结果是正常的，triplet_loss，loss_cls会正常下降，但是多卡的情况下，triplet_loss一直为0.
我定位到fastreid/modeling/losses/triplet_loss.py中的is_pos,is_neg在多卡训练和单卡训练情况下，输出的结果不一样，我怀疑可能是因为多线程，导致的targets发生了变化，从而导致多卡训练情况下triplet_loss变化异常。
![image](https://user-images.githubusercontent.com/17920852/110939947-2037c100-8371-11eb-9a4b-5355fa0ad095.png)

**Comment by L1aoXingyu** [comment]
因为配置文件里面 batch size 默认为 64，每个 id 采样 16 张图，如果你用4卡训练，相当于每张卡上都是同一个 id 的 16 张图，triplet loss 就是0，如果你要多卡训练，需要同步调大 batch size

**Comment by NiuCY** [comment]
多谢您的回答，问题解决了。
